diff --git a/mlir/include/air-c/Registration.h b/mlir/include/air-c/Registration.h
index b7d9e53e..d4a04a22 100644
--- a/mlir/include/air-c/Registration.h
+++ b/mlir/include/air-c/Registration.h
@@ -18,10 +18,10 @@ extern "C" {
 /** Registers all AIR dialects with a context.
  * This is needed before creating IR for these Dialects.
  */
-void airRegisterAllDialects(MlirContext context);
+MLIR_CAPI_EXPORTED void airRegisterAllDialects(MlirContext context);
 
 /** Registers all AIR passes for symbolic access with the global registry. */
-void airRegisterAllPasses();
+MLIR_CAPI_EXPORTED void airRegisterAllPasses();
 
 #ifdef __cplusplus
 }
diff --git a/mlir/include/air-c/Runner.h b/mlir/include/air-c/Runner.h
index de2c744c..18ab5a3d 100644
--- a/mlir/include/air-c/Runner.h
+++ b/mlir/include/air-c/Runner.h
@@ -17,7 +17,7 @@
 extern "C" {
 #endif
 
-void airRunnerRun(MlirModule module, const char *json_file_name,
+MLIR_CAPI_EXPORTED void airRunnerRun(MlirModule module, const char *json_file_name,
                   const char *output_file_name, const char *function,
                   const char *sim_granularity, bool verbose);
 
diff --git a/mlir/include/air-c/Transform.h b/mlir/include/air-c/Transform.h
index 95ca1458..6bf42a27 100644
--- a/mlir/include/air-c/Transform.h
+++ b/mlir/include/air-c/Transform.h
@@ -16,7 +16,7 @@
 extern "C" {
 #endif
 
-void runTransform(MlirModule transform, MlirModule payload);
+MLIR_CAPI_EXPORTED void runTransform(MlirModule transform, MlirModule payload);
 
 #ifdef __cplusplus
 }
diff --git a/mlir/lib/CAPI/CMakeLists.txt b/mlir/lib/CAPI/CMakeLists.txt
index a0457c81..a5669ebc 100644
--- a/mlir/lib/CAPI/CMakeLists.txt
+++ b/mlir/lib/CAPI/CMakeLists.txt
@@ -2,21 +2,13 @@
 # Copyright (C) 2022, Advanced Micro Devices, Inc. All rights reserved.
 # SPDX-License-Identifier: MIT
 
-add_mlir_library(AIRCAPI
+add_mlir_public_c_api_library(AIRCAPI
 Dialects.cpp
 Registration.cpp
 Runner.cpp
 Transform.cpp
 
-DEPENDS
-
-ENABLE_AGGREGATION
-LINK_COMPONENTS
-Core
-
 LINK_LIBS PUBLIC
 AIRDialect
 AIRInitAll
-MLIRIR
-MLIRSupport
 )
\ No newline at end of file
