diff --git a/CMakeLists.txt b/CMakeLists.txt
index b7770b43..6114b1ee 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -73,10 +73,12 @@ include(AddMLIR)
 include(HandleLLVMOptions)
 
 # setup python
-find_package(Python3 COMPONENTS Interpreter Development REQUIRED)
+#find_package(Python3 COMPONENTS Interpreter Development REQUIRED)
 include(MLIRDetectPythonEnv)
+include(AddMLIRPython)
 mlir_detect_pybind11_install()
-find_package(pybind11 2.6 REQUIRED)
+#find_package(pybind11 2.6 REQUIRED)
+mlir_configure_python_dev_packages()
 
 # python build directory
 if(NOT AIR_PYTHON_PACKAGES_DIR)
