diff --git a/tools/air-runner/CMakeLists.txt b/tools/air-runner/CMakeLists.txt
index 1163d027..24e0e84d 100644
--- a/tools/air-runner/CMakeLists.txt
+++ b/tools/air-runner/CMakeLists.txt
@@ -2,7 +2,10 @@
 # Copyright (C) 2022, Advanced Micro Devices, Inc. All rights reserved.
 # SPDX-License-Identifier: MIT
 
-llvm_map_components_to_libnames(llvm_libs support core irreader)
+set(LLVM_LINK_COMPONENTS
+        Core
+        Support
+)
 
 # Now build our tools
 add_llvm_tool(air-runner air-runner.cpp)
@@ -18,6 +21,7 @@ AIRUtil
 AIRConversionPasses
 AIRTransformPasses
 AIRInitAll
+MLIRSupport
 )
 
 target_link_libraries(air-runner PRIVATE ${LIBS})
