diff --git a/bp_common/src/include/bp_common_aviary_pkgdef.svh b/bp_common/src/include/bp_common_aviary_pkgdef.svh
index 9d967175..1abbade7 100644
--- a/bp_common/src/include/bp_common_aviary_pkgdef.svh
+++ b/bp_common/src/include/bp_common_aviary_pkgdef.svh
@@ -7,6 +7,26 @@
   // Default configuration is unicore
   localparam bp_proc_param_s bp_unicore_cfg_p = bp_default_cfg_p;
 
+  localparam bp_proc_param_s bp_unicore_air_override_p =
+    '{vaddr_width  : 44
+      ,paddr_width : 44
+
+      ,icache_sets        : 512
+      ,icache_assoc       : 1
+      ,icache_block_width : 64
+      ,dcache_sets        : 512
+      ,dcache_assoc       : 1
+      ,dcache_block_width : 64
+
+      ,l2_features : (1 << e_cfg_writeback) | (1 << e_cfg_word_tracking)
+
+      ,default : "inv"
+      };
+  `bp_aviary_derive_cfg(bp_unicore_air_cfg_p
+                        ,bp_unicore_air_override_p
+                        ,bp_unicore_cfg_p
+                        );
+
   localparam bp_proc_param_s bp_unicore_megaparrot_override_p =
     '{paddr_width : 56
       ,daddr_width: 55
@@ -470,8 +490,11 @@
   /* verilator lint_off WIDTH */
   parameter bp_proc_param_s [max_cfgs-1:0] all_cfgs_gp =
   {
+    // AIR configurations
+    bp_unicore_air_cfg_p
+
     // L2 extension configurations
-    bp_multicore_4_l2e_cfg_p
+    ,bp_multicore_4_l2e_cfg_p
     ,bp_multicore_2_l2e_cfg_p
     ,bp_multicore_1_l2e_cfg_p
 
@@ -521,8 +544,11 @@
   // This enum MUST be kept up to date with the parameter array above
   typedef enum bit [lg_max_cfgs-1:0]
   {
+    // AIR configurations
+    e_bp_unicore_air_cfg                            = 33
+
     // L2 extension configurations
-    e_bp_multicore_4_l2e_cfg                        = 32
+    ,e_bp_multicore_4_l2e_cfg                       = 32
     ,e_bp_multicore_2_l2e_cfg                       = 31
     ,e_bp_multicore_1_l2e_cfg                       = 30
 
diff --git a/bp_common/src/v/bsg_edge_extend.v b/bp_common/src/v/bsg_edge_extend.v
index e0c012f1..2343dd4c 100644
--- a/bp_common/src/v/bsg_edge_extend.v
+++ b/bp_common/src/v/bsg_edge_extend.v
@@ -11,7 +11,7 @@ module bsg_edge_extend #(parameter `BSG_INV_PARAM(width_p))
    , output logic [width_p-1:0] data_o
    );
 
-`ifdef VERILATOR
+`ifndef SIMULATION
   bsg_deff_reset
    #(.width_p(width_p))
    edge_deff
