diff --git a/axi/v/bp_axi_top.sv b/axi/v/bp_axi_top.sv
index 3647c5c..575634c 100644
--- a/axi/v/bp_axi_top.sv
+++ b/axi/v/bp_axi_top.sv
@@ -35,6 +35,7 @@ module bp_axi_top
   (input                                       clk_i
    , input                                     reset_i
    , input                                     rt_clk_i
+   , input [io_noc_did_width_p-1:0]            did_i
 
    //======================== Outgoing I/O ========================
    , output logic [m_axil_addr_width_p-1:0]    m_axil_awaddr_o
@@ -167,7 +168,7 @@ module bp_axi_top
         ,.reset_i(reset_i)
 
         // Irrelevant for current AXI wrapper
-        ,.my_did_i('0)
+        ,.my_did_i(did_i)
         ,.host_did_i('0)
         ,.my_cord_i('0)
 
